### PR TITLE
Add System.Half

### LINKED ISSUE #23030: Need to add Half to this page
https://docs.microsoft.com/en-us/dotnet/api/system.half?view=net-5.0

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 5796f0df-3d6f-30e1-ed1a-2169dc27dc38
* Version Independent ID: 56cab56d-07c8-885b-ac7a-b4e90e5409df
* Content: [.NET class library overview](https://docs.microsoft.com/en-us/dotnet/standard/class-library-overview#system-namespace)
* Content Source: [docs/standard/class-library-overview.md](https://github.com/dotnet/docs/blob/main/docs/standard/class-library-overview.md)
* Product: **dotnet-fundamentals**
* GitHub Login: @gewarren
* Microsoft Alias: **gewarren**

### PR DESCRIPTION
**Pull request by gewarren**
Fixes #23030. (Hide whitespace changes.)

See https://github.com/dotnet/runtime/pull/37630 and https://github.com/dotnet/runtime/issues/936.

[Internal preview](https://review.docs.microsoft.com/en-us/dotnet/standard/numerics?branch=pr-en-us-23855#floating-point-types).

### CODE REVIEW

File: docs/standard/numerics.md
Comment (gewarren) [inline docs/standard/numerics.md:55]: @pgovind Can you help me add the approximate range and precision values?

File: docs/standard/base-types/standard-numeric-format-strings.md
Comment (gewarren) [inline docs/standard/base-types/standard-numeric-format-strings.md:347]: @pgovind Is there a special format specifier to round-trip a Half value like there is for Double and Single (see line 288)?

Comment (gewarren) [reply docs/standard/base-types/standard-numeric-format-strings.md:347]: Also, I noticed that Half has properties whereas Single and Double have fields. Is that intentional?

File: docs/standard/numerics.md
Comment (IEvangelist) [reply docs/standard/numerics.md]: I was going to suggest add the appropriate range too:

- [min](https://review.docs.microsoft.com/en-us/dotnet/api/system.half.minvalue?view=net-5.0&branch=main#System_Half_MinValue)
- [max](https://review.docs.microsoft.com/en-us/dotnet/api/system.half.maxvalue?view=net-5.0&branch=main#System_Half_MaxValue)

File: docs/standard/base-types/custom-numeric-format-strings.md
Comment (adegeo) [inline docs/standard/base-types/custom-numeric-format-strings.md:257]: missing space
```suggestion
Regardless of the format string, if the value of a <xref:System.Half>, <xref:System.Single>, or <xref:System.Double> floating-point type is positive infinity, negative infinity, or not a number (NaN), the formatted string is the value of the respective <xref:System.Globalization.NumberFormatInfo.PositiveInfinitySymbol%2A>, <xref:System.Globalization.NumberFormatInfo.NegativeInfinitySymbol%2A>, or <xref:System.Globalization.NumberFormatInfo.NaNSymbol%2A> property specified by the currently applicable <xref:System.Globalization.NumberFormatInfo> object.
```

File: docs/standard/language-independence.md
Comment (adegeo) [inline docs/standard/language-independence.md:298]: It may be worth pointing out what is said in the API docs:

>An IEEE 754 compliant `float16` type.

File: docs/standard/language-independence-and-language-independent-components.md
Comment (adegeo) [inline docs/standard/language-independence-and-language-independent-components.md:175]: Same note here as before

File: docs/standard/base-types/standard-numeric-format-strings.md
Comment (tannergooding) [reply docs/standard/base-types/standard-numeric-format-strings.md:347]: This is intentional. `double` and `single` are runtime primitives and have special encoding in metadata for constants. `Half` is functionally a "user-defined" type and so cannot declare constants today.

File: docs/standard/class-library-overview.md
Comment (tannergooding) [inline docs/standard/class-library-overview.md:95]: This is unrelated to the current PR, but `IntPtr` and `UIntPtr` now have built-in types on C#, these are `nint` and `nuint`, respectively.

File: docs/standard/class-library-overview.md
Comment (tannergooding) [inline docs/standard/class-library-overview.md:90]: Should `Half` be listed here as well? It isn't currently a language primitive but it is a "fundamental type" in that it is part of the ABI for various platforms and may become a language primitive in the future.

The same would be true if we added `Int128` and `UInt128` types in the future, for example.

File: docs/standard/language-independence-and-language-independent-components.md
Comment (tannergooding) [reply docs/standard/language-independence-and-language-independent-components.md:175]: The formal IEEE 754 name is `binary16` for `half`, `binary32` for single, and `binary64` for `double`.

File: docs/standard/numerics.md
Comment (tannergooding) [inline docs/standard/numerics.md:36]: Not related to this PR, but should `IntPtr` (`nint`) and `UIntPtr` (`nuint`) be listed here as well?

They are fundamental integer types, but have different ranges when running on 32-bit vs 64-bit

File: docs/standard/numerics.md
Comment (tannergooding) [inline docs/standard/numerics.md:55]: The exact range for `Half` is `+/-65504`.

The number of digits, mirroring what is used in the other columns is `~3-5 digits`. However, that column is very misleading in my opinion and may give users a false sense that any value with 6 significant digits can be exactly represented.

In practice, the lower limit here is that if a string with at most `n` significant digits (`3`, `6`, `15` for `half`, `float`, and `double`, respectively) is converted to a floating-point value and then back to a string with the same number of digits, the strings should match. That is, `float.Parse(s).ToString("G6") == s` should be `true` for any `s` that contains no more than `6` significant digits.
* This isn't guaranteed by the IEEE 754 spec and may depend on the formatting algorithm used

While the upper limit here is that if a floating-point value is converted to a string with that many digits (`5`, `9`, `17` for `half`, `float`, and `double`, respectively), and then converted back to a floating-point value, the result should match the original floating-point value. That is, `float.Parse(x.ToString("G9")).Equals(x)` should be `true` for any `float`.
* This is guaranteed by the IEEE 754 spec

File: docs/standard/language-independence-and-language-independent-components.md
Comment (adegeo) [reply docs/standard/language-independence-and-language-independent-components.md:175]: @tannergooding we should probably update the API docs, as it says float16: https://docs.microsoft.com/en-us/dotnet/api/system.half?view=net-5.0

File: docs/standard/base-types/standard-numeric-format-strings.md
Comment (pgovind) [reply docs/standard/base-types/standard-numeric-format-strings.md:347]: > Is there a special format specifier to round-trip a Half value like there is for Double and Single (see line 288)

`G5` for Half

**Review by pgovind — APPROVED** [review]
Other reviews have called out everything I noticed too. Approving assuming the comments will be resolved! Overall LGTM!

File: docs/standard/base-types/standard-numeric-format-strings.md
Comment (tannergooding) [reply docs/standard/base-types/standard-numeric-format-strings.md:347]: Since `Half` is only supported on .NET Core, `R` is the right thing to use here and will always do the correct thing.

Comment (tannergooding) [reply docs/standard/base-types/standard-numeric-format-strings.md:347]: We only need to list `G9` and `G17` for float/double because `R` had bugs on full framework.

File: docs/standard/numerics.md
Comment (pgovind) [reply docs/standard/numerics.md]: Called out elsewhere already. Just for complete-ness: The range is `+/-65504`

File: docs/standard/base-types/standard-numeric-format-strings.md
Comment (pgovind) [reply docs/standard/base-types/standard-numeric-format-strings.md:347]: Ah that explains why. I was wondering what cases `R` gave wrong results for.

File: docs/standard/numerics.md
Comment (gewarren) [reply docs/standard/numerics.md]: Thanks. For System.Single, this table shows the low end of the range as 1.5x10<sup>-45</sup>. But the [API ref docs show the MinValue](https://docs.microsoft.com/en-us/dotnet/api/system.single.minvalue?view=net-5.0) as -3.40282347E+38. Which one is correct?

Comment (pgovind) [reply docs/standard/numerics.md]: `MinValue` is `-3.40282347E+38`. However, there is some nuance here about normal and subnormal numbers. @tannergooding, we should decide what we want to docs to state the range as (applies to Double too)

Comment (tannergooding) [reply docs/standard/numerics.md]: One is giving the minimum value and the other is giving the smallest representable magnitude (that is the smallest +/- value which is not zero)

Comment (pgovind) [reply docs/standard/numerics.md]: Indeed, but the way the range is specified in the docs right now is misleading IMO. Does it make more sense here to just state the range as `+-3.40282347E+38`. The documentation for `Epsilon` at https://docs.microsoft.com/en-us/dotnet/api/system.single.epsilon?view=net-5.0 is quite good already.

Comment (tannergooding) [reply docs/standard/numerics.md]: > Indeed, but the way the range is specified in the docs right now is misleading IMO

Agreed, it can be confusing depending on how you read it and how familiar you are with IEEE 754 floating-point.

Comment (pgovind) [reply docs/standard/numerics.md]: Cool. I think we should just state the range here as `+-3.4E+38` for Single, `+-1.7E+308` for Double and `+-7.9E+28` for Decimal.

File: docs/standard/class-library-overview.md
Comment (gewarren) [reply docs/standard/class-library-overview.md:90]: It's there already on line 89.

File: docs/standard/numerics.md
Comment (gewarren) [reply docs/standard/numerics.md:55]: I will plan to remove that column then. Thanks!

File: docs/standard/class-library-overview.md
Comment (tannergooding) [reply docs/standard/class-library-overview.md:90]: Completely missed it 😅

**Comment by gewarren** [comment]
Thanks for the great reviews!

**Pull request merged**
